File: 0104-maximum-depth-of-binary-tree/0104-maximum-depth-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        def rec(node, depth):
            if not node:
                return depth - 1
            if not node.left and not node.right:
                return depth
            left, right = depth, depth
            if node.left:
                left = rec(node.left, depth + 1)
            if node.right:
                right = rec(node.right, depth + 1)
            return max(left, right)
        return rec(root, 1)

File: 0104-maximum-depth-of-binary-tree/0104-maximum-depth-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int recurse(TreeNode node, int depth){
        if(node == null)
            return depth - 1;
        int left = recurse(node.left, depth + 1);
        int right = recurse(node.right, depth + 1);
        return Math.max(left, right);
    }
    public int maxDepth(TreeNode root) {
        return recurse(root, 1);
    }
}